docs/specs/instrumentation.md

```diff
@@ -8,34 +8,145 @@ interpreted as described in [RFC 2119].
+  * `name` - is one or more words that describes the measurement (e.g., `memory_rss`, `requests`)
+  * `unit` - MUST be a single [base unit] in plural form, if applicable (e.g., `seconds`, `bytes`)
   * Counters MUST end with `total` (e.g., `disk_written_bytes_total`, `http_requests_total`)
 * SHOULD be broad in purpose and use use tags to differentiate characteristics of the measurement (e.g., `host_cpu_seconds_total{cpu="0",mode="idle"}`)
 
+## Emission
+
+### Batching
+
+For performance reasons, as demonstrated in pull request #8383],
+instrumentation SHOULD be batched whenever possible:
+
+* Vector process batches of events ([RFC 9480]) and telemtry SHOULD emit for
+  the entire batch, not each individual event.
+
+### Events
+
+Instrumentation SHOULD be event-driven ([RFC 2064]), where individual events
+serve as the vehicle for internal telemtry, driving the emission of metrics
+and logs. This organizes Vector's telemetry, making it easier to manage and 
+catalogue. On rare occassions, metrics and logs can emit directly but MUST be
+reserved for ocassions where it is impossible to emit Vector's events. For
+example, emitting metrics in a library that cannot import Vector's events.
+
+#### Error
+
+An `<Name>Error` event MUST be emitted when an error occurs. Errors that are
+retriable and possible recoverable MUST be distinguished from errors that are
+not:
+
+* Properties
+  * `retriable` - If the operation that produced the error is retriable. This
+    indicates that the error operation might recover and determines if this
+    error should be a warning or an error when emitting logs and metrics. For
+    example, a failed HTTP request that will be retried.
+  * `error_code` - An error code for the failure, if applicable.
+    * SHOULD only be specified if it adds additional information beyond
+      `error_type`.
+    * The values for `error_code` for a given error event MUST be a bounded set
+      with relatively low cardinality because it will be used as a metric tag.
+      Examples would be syscall error code. Examples of values that should not
+      be used are raw error messages from `serde` as these are highly variable
+      depending on the input. Instead, these errors should be converted to an
+      error code like `invalid_json`.
+  * `error_type` - The type of error condition. MUST be one of the types listed
+    in the `error_type` enum list in the cue docs.
+  * `stage` - The stage at which the error occurred. MUST be one of `receiving`,
+    `processing`, or `sending`.
+  * If any of the above properties are implicit to the specific error
+    type, they MAY be omitted from being represented explicitly in the
+    event fields. However, they MUST still be included in the emitted
+    logs and metrics, as specified below, as if they were present.
+* Metrics
+  * MUST include the defined properties as tags.
+  * If `retriable` is `true`, MUST increment `<namespace>_warnings_total` metric.
```

Comment:
I'd suggest we drop this metric until we learn more.

Reply:
I would agree: it's hard to gauge what signal this metric would provide as-is.

Part of me would almost rather just always emit a log/metric for all errors -- because they _are_ errors -- and depend on discarded events to figure out how many actual events have been lost, if any.

Reply:
Resolved in https://github.com/vectordotdev/vector/pull/12822/commits/417f2667c3a27eb099c7c332a82f689e5aad25e8